sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryCredentialsPolicy.cs

```diff
@@ -0,0 +1,131 @@
+    /// Response Body: { acrRefreshToken }
+    ///
+    /// Step 4: POST /api/oauth2/token
+    /// Request Body: { acrRefreshToken, scope, grant-type }
+    /// Response Body: { acrAccessToken }
+    ///
+    /// Step 5: GET /api/v1/acr/repositories
+    /// Request Header: { Bearer acrTokenAccess }
+    /// </summary>
+    internal class ContainerRegistryCredentialsPolicy : BearerTokenChallengeAuthenticationPolicy
+    {
+        private readonly RefreshTokensRestClient _exchangeRestClient;
+        private readonly AccessTokensRestClient _tokenRestClient;
+
+        public ContainerRegistryCredentialsPolicy(TokenCredential credential, string aadScope, RefreshTokensRestClient exchangeRestClient, AccessTokensRestClient tokenRestClient)
+            : base(credential, aadScope)
+        {
+            Argument.AssertNotNull(credential, nameof(credential));
+            Argument.AssertNotNull(aadScope, nameof(aadScope));
+
+            _exchangeRestClient = exchangeRestClient;
+            _tokenRestClient = tokenRestClient;
+        }
+
+        protected override async ValueTask<bool> AuthenticateRequestOnChallengeAsync(HttpMessage message, bool async)
+        {
+            // Once we're here, we've completed Step 1.
+
+            // Step 2: Parse challenge string to retrieve serviceName and scope, where scope is the ACR Scope
+            var service = AuthorizationChallengeParser.GetChallengeParameterFromResponse(message.Response, "Bearer", "service");
+            var scope = AuthorizationChallengeParser.GetChallengeParameterFromResponse(message.Response, "Bearer", "scope");
+
+            string acrAccessToken = string.Empty;
+            if (async)
+            {
+                // Step 3: Exchange AAD Access Token for ACR Refresh Token
+                string acrRefreshToken = await ExchangeAadAccessTokenForAcrRefreshTokenAsync(message, service, true).ConfigureAwait(false);
+
+                // Step 4: Send in acrRefreshToken and get back acrAccessToken
+                acrAccessToken = await ExchangeAcrRefreshTokenForAcrAccessTokenAsync(acrRefreshToken, service, scope, true).ConfigureAwait(false);
+            }
+            else
+            {
+                // Step 3: Exchange AAD Access Token for ACR Refresh Token
+                string acrRefreshToken = ExchangeAadAccessTokenForAcrRefreshTokenAsync(message, service, false).EnsureCompleted();
+
+                // Step 4: Send in acrRefreshToken and get back acrAccessToken
+                acrAccessToken = ExchangeAcrRefreshTokenForAcrAccessTokenAsync(acrRefreshToken, service, scope, false).EnsureCompleted();
+            }
+
+            // Step 5 - Authorize Request.  Note, we don't use SetAuthorizationHeader here, because it
+            // sets an AAD access token header, and at this point we're done with AAD and using an ACR access token.
+            message.Request.Headers.SetValue(HttpHeader.Names.Authorization, $"Bearer {acrAccessToken}");
+
+            return true;
```

Comment:
Is there any explicit failure scenario where we'd want to return `false`?

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/19791